exchanges/order/limits.go

```diff
@@ -0,0 +1,378 @@
+		limit.marketStepIncrementSize = levels[x].MarketStepSize
+		limit.maxTotalOrders = levels[x].MaxTotalOrders
+		limit.maxAlgoOrders = levels[x].MaxAlgoOrders
+		limit.m.Unlock()
+	}
+	return nil
+}
+
+// GetOrderExecutionLimits returns the exchange limit parameters for a currency
+func (e *ExecutionLimits) GetOrderExecutionLimits(a asset.Item, cp currency.Pair) (*Limits, error) {
+	e.mtx.RLock()
+	defer e.mtx.RUnlock()
+
+	if e.m == nil {
+		return nil, ErrExchangeLimitNotLoaded
+	}
+
+	m1, ok := e.m[a]
+	if !ok {
+		return nil, errExchangeLimitAsset
+	}
+
+	m2, ok := m1[cp.Base]
+	if !ok {
+		return nil, errExchangeLimitBase
+	}
+
+	limit, ok := m2[cp.Quote]
+	if !ok {
+		return nil, errExchangeLimitQuote
+	}
+
+	return limit, nil
+}
+
+// CheckOrderExecutionLimits checks to see if the price and amount conforms with
+// exchange level order execution limits
+func (e *ExecutionLimits) CheckOrderExecutionLimits(a asset.Item, cp currency.Pair, price, amount float64, orderType Type) error {
+	e.mtx.RLock()
+	defer e.mtx.RUnlock()
+
+	if e.m == nil {
+		// No exchange limits loaded so we can nil this
+		return nil
+	}
+
+	m1, ok := e.m[a]
+	if !ok {
+		return errCannotValidateAsset
+	}
+
+	m2, ok := m1[cp.Base]
+	if !ok {
+		return errCannotValidateBaseCurrency
+	}
+
+	limit, ok := m2[cp.Quote]
+	if !ok {
+		return errCannotValidateQuoteCurrency
+	}
+
+	err := limit.Conforms(price, amount, orderType)
+	if err != nil {
+		return fmt.Errorf("%w for %s %s", err, a, cp)
+	}
+
+	return nil
+}
+
+// Limits defines total limit values for an associated currency to be checked
+// before execution on an exchange
+type Limits struct {
+	minPrice                float64
+	maxPrice                float64
+	stepIncrementSizePrice  float64
+	minAmount               float64
+	maxAmount               float64
+	stepIncrementSizeAmount float64
+	minNotional             float64
+	multiplierUp            float64
+	multiplierDown          float64
+	averagePriceMinutes     int64
+	maxIcebergParts         int64
+	marketMinQty            float64
+	marketMaxQty            float64
+	marketStepIncrementSize float64
+	maxTotalOrders          int64
+	maxAlgoOrders           int64
+	m                       sync.RWMutex
+}
+
+// Conforms checks outbound parameters
+func (l *Limits) Conforms(price, amount float64, orderType Type) error {
+	if l == nil {
+		// For when we return a nil pointer we can assume there's nothing to
+		// check
+		return nil
+	}
+
+	l.m.RLock()
+	defer l.m.RUnlock()
+	if l.minAmount != 0 && amount < l.minAmount {
+		return fmt.Errorf("%w min: %.8f supplied %.8f",
+			ErrAmountBelowMin,
+			l.minAmount,
+			amount)
+	}
+	if l.maxAmount != 0 && amount > l.maxAmount {
+		return fmt.Errorf("%w min: %.8f supplied %.8f",
+			ErrAmountExceedsMax,
+			l.maxAmount,
+			amount)
+	}
+	if l.stepIncrementSizeAmount != 0 {
+		dAmount := decimal.NewFromFloat(amount)
+		dStep := decimal.NewFromFloat(l.stepIncrementSizeAmount)
+		if !dAmount.Mod(dStep).IsZero() {
```

Comment:
Yeah, this was a weird one, I'm pretty sure we can cancel out that because we check for if the quantity is below the minQty by the time it hits this check. If you minus the minQty from the equation it just reduces your floor it doesn't look like it affects the mod equation at all. Will run another check to make sure.